### PR TITLE
Add support of `srcset` to `yii\helpers\Html::img()`

### PR DESCRIPTION
**Pull request by Kolyunya**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #10834 

Not sure if we should distinguish between `src` and `srcset` passed as the first argument using `yii\helpers\ArrayHelper::isAssociative()` or just pass `srcset` in `options` while passing `null` to `$src`.

If we choose the former it will hit the performance of each and every call to `Html::img()` in all existing and future applications. We also won't be able to support non-assiciative `srcsets` like following. Missing descriptor defaults to `1x` according to the [docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). Not sure if we should support this though.
```php
$srcset = [
    'base-url',
    '2x' => 'other-url',
];
```

If we choose the latter the method invocation will look somewhat _weird_ but I think we can live with it:
```php
<?= Html::img(null, ['srcset' => [
    '1x' => 'foo',
    '2x' => 'bar',
]]); ?>
```

### CODE REVIEW

**Comment by samdark** [comment]
There's another implementation option — a separate method.

**Comment by Kolyunya** [comment]
@samdark speaking of semantics this is still and `img` tag but with other attributes. I'd personally stick to the option of passing `srcset` in `options`.

If we decide to implement a separate tag what name should it have? Maybe `imgSet`? Or `imgSrcset`? Or `srcsetImg`?

**Comment by samdark** [comment]
Either one.

## 1\.  `imgSrcset` (separate method)

- Cleaner.
- Less complicated.
- Method name doesn't precisely match tag name.


## 2\. `srcset` in options

- Still perfect tag name-method name match.
- Weird syntax.
- Complicated logic.

**Comment by rob006** [comment]
@samdark What is weird in this syntax:

```php
<?= Html::img('img.png', ['srcset' => [
    '1x' => 'img-1x.png',
    '2x' => 'img-2x.png',
]]); ?>
```
?

**Comment by Kolyunya** [comment]
@samdark it seems like this method must support both `src` and `srcset` attributes at the same time. Updated the PR accordingly.

**Comment by rob006** [comment]
There is also `sizes`attribute for `img` tag, but not sure if it needs special syntax.

File: framework/helpers/BaseHtml.php [hunk 1]
```diff
@@ -434,7 +434,18 @@ public static function mailto($text, $email = null, $options = [])
      */
     public static function img($src, $options = [])
     {
-        $options['src'] = Url::to($src);
+        if (isset($src)) {
+            $options['src'] = Url::to($src);
+        }
```
Comment (rob006) [inline framework/helpers/BaseHtml.php]: AFAIK `src` attribute is required for `img` tag. Why it is optional here?

**Comment by Kolyunya** [comment]
@rob006 seems like there is not much we can do for `sizes`. Example from [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#Example_4_Using_the_srcset_and_sizes_attributes).
```
<img src="clock-demo-thumb-200.png" 
      alt="Clock" 
      srcset="clock-demo-thumb-200.png 200w,
          clock-demo-thumb-400.png 400w"
      sizes="(min-width: 600px) 200px, 50vw">
```

File: framework/helpers/BaseHtml.php (hunk 1 above)
Comment (Kolyunya) [inline framework/helpers/BaseHtml.php]: @rob006 that is [true](https://www.w3.org/TR/html5/embedded-content-0.html#the-img-element).

> The src attribute must be present, and must contain a valid non-empty URL potentially surrounded by spaces referencing a non-interactive, optionally animated, image resource that is neither paged nor scripted.

**Comment by rob006** [comment]
@Kolyunya Using array for `sizes` list could improve readability - when each option is in its own line it is harder to miss something.

File: framework/helpers/BaseHtml.php (hunk 1 above)
Comment (Kolyunya) [inline framework/helpers/BaseHtml.php]: @rob006 fixed.

**Comment by Kolyunya** [comment]
Added support for passing `srcset` as a `string` for BC.

**Comment by samdark** [comment]
@rob006 it is fine.

**Review by samdark — APPROVED** [review]

File: tests/framework/helpers/HtmlTest.php [hunk 2]
```diff
@@ -137,6 +137,30 @@ public function testImg()
         $this->assertEquals('<img src="/example" alt="">', Html::img('/example'));
         $this->assertEquals('<img src="/test" alt="">', Html::img(''));
         $this->assertEquals('<img src="/example" width="10" alt="something">', Html::img('/example', ['alt' => 'something', 'width' => 10]));
+        $this->assertEquals('<img src="/base-url" alt="" srcset="/example-100w 100w,/example-500w 500w,/example-1500w 1500w">', Html::img('/base-url', [
+            'srcset' => [
+                '100w' => '/example-100w',
+                '500w' => '/example-500w',
+                '1500w' => '/example-1500w',
```
Comment (dynasource) [inline tests/framework/helpers/HtmlTest.php]: are three config options really required for this test?

File: tests/framework/helpers/HtmlTest.php [hunk 3]
```diff
@@ -137,6 +137,30 @@ public function testImg()
...
+            ],
+        ]));
+        $this->assertEquals('<img src="/base-url" alt="" srcset="/example-1x 1x,/example-2x 2x,/example-3x 3x">', Html::img('/base-url', [
+            'srcset' => [
+                '1x' => '/example-1x',
+                '2x' => '/example-2x',
+                '3x' => '/example-3x',
```
Comment (dynasource) [inline tests/framework/helpers/HtmlTest.php]: same

File: tests/framework/helpers/HtmlTest.php [hunk 4]
```diff
@@ -137,6 +137,30 @@ public function testImg()
...
+        $this->assertEquals('<img src="/base-url" alt="" srcset="/example-1.42x 1.42x,/example-2.0x 2.0x,/example-3.999x 3.999x">', Html::img('/base-url', [
+            'srcset' => [
+                '1.42x' => '/example-1.42x',
+                '2.0x' => '/example-2.0x',
+                '3.999x' => '/example-3.999x',
+            ],
+        ]));
+        $this->assertEquals('<img src="/base-url" alt="" srcset="/example-1x 1x,/example-2x 2x,/example-3x 3x">', Html::img('/base-url', [
+            'srcset' => '/example-1x 1x,/example-2x 2x,/example-3x 3x',
+        ]));
```
Comment (dynasource) [inline tests/framework/helpers/HtmlTest.php]: the expansion of this unit test shows the opportunity for a PhpUnit dataprovider to provide:
- code reuse
- clear overview on input data
- easier expansion

**Review by dynasource — CHANGES_REQUESTED** [review]

File: tests/framework/helpers/HtmlTest.php (hunk 2 above)
Comment (Kolyunya) [inline tests/framework/helpers/HtmlTest.php]: Yes, why not? I think we should have at least the tests with 0, 1 and 3 options.

Comment (dynasource) [inline tests/framework/helpers/HtmlTest.php]: aren't 2 options enough to make your point in the assertion?

**Comment by Kolyunya** [comment]
@dynasource I refactored the tests to use data provider as you advised.

**Comment by dynasource** [comment]
looks good. Thanks for adding the dataProvder. I am still trying to grasp why so many 'srcset' options are required for the assertions.

**Comment by Kolyunya** [comment]
@dynasource zero and one are corner-cases; three and five are regular cases with `w`, `int` and `float` descriptors.

**Comment by dynasource** [comment]
I see. My goal is to have these tests as short as possible (for maintainability purposes) without impacting the coverage of potential examples in practice.

**Review by dynasource — APPROVED** [review]

**Comment by Kolyunya** [comment]
```php
     * @since 2.0.12 It is possible to pass the "srcset" option as an array which keys are
     * descriptors and values are URLs. All URLs will be processed by [[Url::to()]].
````
Added a note the method documentation.

**Comment by Kolyunya** [comment]
Can we assign this PR to `2.0.12`? Is there anything else that should be done here?

**Comment by samdark** [comment]
It's fine. Merged. Thank you!

File: framework/helpers/BaseHtml.php [hunk 5]
```diff
@@ -430,11 +430,22 @@ public static function mailto($text, $email = null, $options = [])
      * the attributes of the resulting tag. The values will be HTML-encoded using [[encode()]].
      * If a value is null, the corresponding attribute will not be rendered.
      * See [[renderTagAttributes()]] for details on how attributes are being rendered.
+     * @since 2.0.12 It is possible to pass the "srcset" option as an array which keys are
+     * descriptors and values are URLs. All URLs will be processed by [[Url::to()]].
```
Comment (cebe) [inline framework/helpers/BaseHtml.php]: this should not be an `@since` tag as that would apply to the method itself, which is there already longer. fixed by e12dd89

Comment (Kolyunya) [inline framework/helpers/BaseHtml.php]: @cebe wouldn't it be better to add a separate `@since` for when this method was introduced and leave the `@since 2.0.12` for that patch?

Comment (Kolyunya) [inline framework/helpers/BaseHtml.php]: @cebe you can have multiple `@since` tags according to the [docs](https://phpdoc.org/docs/latest/references/phpdoc/tags/since.html) to specify what happened in which patch. Having a single `@since` doesn't mean that this element was introduced in this patch.

Comment (cebe) [inline framework/helpers/BaseHtml.php]: Interesting, but this is not how our apidoc uses it :)
We currently use `@since` only to indicate when a class or method has been introduced.
Seems this is some kind of changelog for the method, which is a nice idea as we could probably use it to generate more detailed changelogs and provide information on which methods have changed in a release. Its not how it works currently though, so if we start doing this we need to adjust it in other places also and apply this approach to the whole framework.

Comment (rob006) [inline framework/helpers/BaseHtml.php]: Created issue for that: https://github.com/yiisoft/yii2-apidoc/issues/140
I think it is worth to have at some point, but it is not a priority right now.

Comment (cebe) [inline framework/helpers/BaseHtml.php]: Thanks!
